rfcs/react-components/convergence/component-transitions-on-mount-or-unmount.md

```diff
@@ -0,0 +1,125 @@
+# RFC: Component CSS Transitions on mount/unmount
+
+<!--
+An RFC can be anything. A question, a suggestion, a plan. The purpose of this template is to give some structure to help folks write successful RFCs. However, don't feel constrained by this template; use your best judgement.
+
+Tips for writing a successful RFC:
+
+- Simple plain words that make your point, fancy words obfuscate
+- Try to stay concise, but don't gloss over important details
+- Try to write a neutral problem statement, not one that motivates your desired solution
+- Remember, "Writing is thinking". It's natural to realize new ideas while writing your proposal
+-->
+
+---
+
+@marcosmoura
+
+## Summary
+
+This RFC describes how CSS transitions in Fluent UI components can be applied.
+
+## Background
+
+Currently it is not possible to add CSS transition to components on mount/unmount. A usage of this pattern would be transitions to components that has a open/close toggle, such as Drawer, Dialog and more.
+
+## Problem statement
+
+In order to display that a content is showing or hiding from screen, CSS transitions should be applied. Due to React's way of rendering, there is no builtin support for this, therefore we need to create a solution that can provide with the status of component on screen.
+
+## Detailed Design or Proposal
+
+### A `useAnimationState` hook based solution
+
+To differentiate the animation state of a component, a hook could be created and be part of the react-utilities. Note that every time a status change (from not visible to entering for example) the component will rerender. A draft implementation of the hook: [useAnimationState](https://github.com/marcosmoura/fluentui/blob/marcosmoura/feat/drawer-component/packages/react-components/react-drawer/src/components/Drawer/useAnimationState.ts)
+
+Usage in the component:
+
+```ts
+import * as React from 'react';
+import { getNativeElementProps, useAnimationState } from '@fluentui/react-utilities';
+
+import type { SampleProps, SampleState } from './Sample.types';
+
+export const useSample_unstable = ({ open = false }: SampleProps, ref: React.Ref<HTMLElement>): SampleState => {
+  const { visible, mounted, animating, entering, exiting } = useAnimationState(open, {
+    enterDuration: 200,
+    exitDuration: 250,
+  });
+
+  return {
+    components: {
+      root: 'div',
+    },
+
+    root: getNativeElementProps('div'),
+
+    visible,
+    mounted,
+    animating,
```

Comment:
What's the reason for having `animating` and `mounted` states? I don't see them being used in any of the examples. If there are are use cases for them it would be worth having in this RFC

Reply:
That's fair. I will add a description for each value returned.
`mounted`: The mounted is the first tick before `visible` is true and last after `visible` is false.
`animating`: Is a way for us to detect the element should be animating and therefore we can apply animations or transitions to an element, without have them by default. This is useful to when the animation of entering and exiting is the same.

Reply:
Can you please create code (sandbox?) examples for those properties? It would be nice to have reference examples of how to use those properties. We should have scenarios to use these properties in if we want to support them

Reply:
This is now updated.